<a href="https://colab.research.google.com/github/623021000-1thaksina/data_viz_2022/blob/main/Pandas101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd  #pandas ใช้กับข้อมูลที่เป็นตาราง/เอกเซล

.csv = Comma Separated Values

data.go.th

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os #os จัดการไฟล์ในคอมพิวเตอร์ 

In [4]:
path = '/content/drive/My Drive/dataviz_2021_data' 

In [5]:
covid_file_path = os.path.join(path,'pm-20-jan-2021.csv') 
print(covid_file_path)

/content/drive/My Drive/dataviz_2021_data/pm-20-jan-2021.csv


In [6]:
a = path + '/pm-18-jan-2021.csv' 
print(a)

/content/drive/My Drive/dataviz_2021_data/pm-18-jan-2021.csv


## load data to memory

In [7]:
data_covid = pd.read_csv(covid_file_path) #csv เพื่อมาใส่ใน data covid
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

NAN คือไม่มีค่าตัวเลข

In [8]:
data_covid.head(20) #.head ใช้ดูหัวตาราง 

No.  ...                                               risk
0     1  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
1     2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2     3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
3     4  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
4     5  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
5     6  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
6     7  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7     8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
8     9  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
9    10  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
10   11  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11   12  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12   13  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
13   14  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
14   15  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
15   16  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
16   17  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
17   18  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
18   19  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
19   20  ...                      คนไทยเดินทางกลับจากต่างประเทศ

[20 rows x 10 columns]

In [9]:
data_covid.tail()

No.  ...                                         risk
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5 rows x 10 columns]

In [10]:
data_covid.shape #shape ใช้แสดงว่ามีกี่ row กี่ colum

(12653, 10)

## ชี้ค่าในตาราง

#### basic

In [11]:
data_covid['province_of_onset'] 

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [12]:
data_covid['risk'][12649] #คอลัม risk แถวที่ 12649

'สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้'

In [13]:
data_covid['province_of_onset'][4] #ชี้ว่าคอลัมไหน แถวที่ 4

'กรุงเทพมหานคร'

#### .iloc

In [14]:
data_covid.iloc[4,7] #แถวที่ 4 หลักที่ 7

'กรุงเทพมหานคร'

## Table slicing

#### ตัดมาเฉพาะ column ที่ต้องการ

In [15]:
smaller_table = data_covid[['announce_date','province_of_onset','risk']]
smaller_table

announce_date  ...                                         risk
0         1/12/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1         1/17/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4         1/24/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...             ...  ...                                          ...
12648     1/20/2021  ...                             State Quarantine
12649     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651     1/20/2021  ...                                Cluster ระยอง
12652     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 3 columns]

#### ตัดเฉพาะ rows ที่ต้องการ

##### แบบง่าย แต่ใช้การอะไรไม่ค่อยจะได้ (มองเป็น array แล้วตัดแบบ array)

In [16]:
data_covid.iloc[1:5,:]

No. announce_date  ... district_of_onset                              risk
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 10 columns]

แบบ advanced แต่ใช้ประโยชน์ได้เยอะกว่า (ตัดแบบใช้ logic query)

$pddf$[list_of_logical_values]$ คือการเลือก row ที่ logical values เป็น True

In [17]:
data_covid['province_of_onset']

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [18]:
data_covid['province_of_onset']=='ขอนแก่น' #ตารางไหนที่ไม่ใช่ขอนแก่นจะขึ้น Flase

0        False
1        False
2        False
3        False
4        False
         ...  
12648    False
12649    False
12650    False
12651    False
12652    False
Name: province_of_onset, Length: 12653, dtype: bool

In [19]:
data_covid[data_covid['province_of_onset']=='ขอนแก่น']

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

In [20]:
data_covid[data_covid['age']=='70'] # == ใช้ตรวจสอบค่าที่ตรงกันถ้าไม่ตรงกันก็จะออกมาเป็น false ถ้าตรงกันจะออกมาเป็น True

Empty DataFrame
Columns: [No., announce_date, notification_date, sex, age, nationality, province_of_isolation, province_of_onset, district_of_onset, risk]
Index: []

การชี้ค่าในตารางด้วย data_covid[***x***] ค่า ***x*** สามารถมีค่าที่ใส่เข้าไปได้ 2 แบบ


1.   ชื่อ column
2.   ใส่ rows ที่ต้องการ (ใส่ list ที่มี***ขนาดเท่ากับจำนวนแถว*** แล้วบอกว่าเอา แถวไหนบ้างโดยใช้ logical expression (```True```/```False```))

In [21]:
eight_rows_covid = data_covid.iloc[:8,3:] # ตารางตัวอย่าง
eight_rows_covid

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
7  หญิง  73.0  ...            หัวหิน  คนต่างชาติเดินทางมาจากต่างประเทศ

[8 rows x 7 columns]

In [22]:
eight_rows_covid[[True,False,True,True,False,True,False,False]]   #2.  ใส่ list ที่ มีค่า logic ของ แต่ละ row โดยที่ ค่า True คือ เลือก ค่า False คือไม่เลือก

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 7 columns]

In [23]:
eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร'  #สร้าง list ของ logical expression ค่าไหนในตารางตัวอย่างเป็นกรุงเทพบ้าง?

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: province_of_onset, dtype: bool

In [24]:
eight_rows_covid[eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร']  #เราเอา 2 บรรทัดบนมาบวกกัน

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[6 rows x 7 columns]

In [25]:
eight_rows_covid['district_of_onset'] == 'เมือง'

0    False
1    False
2     True
3    False
4    False
5    False
6    False
7    False
Name: district_of_onset, dtype: bool

In [26]:
import numpy as np

$.isnull()$ ใช้ตรวจสอบว่าเป็น NAN หรือไม่

In [27]:
eight_rows_covid['district_of_onset'].isnull()

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: district_of_onset, dtype: bool

Quiz (ให้ใช้เฉพาะที่อาจารย์สอนไปแล้วในวิชานี้)

1.   คำนวณ อายุเฉลี่ย ของ**ผู้ชาย** ในข้อมูลทั้งหมด
2.   คำนวณ อายุเฉลี่ย ของ**ผู้ชาย** ในจังหวัดขอนแก่น
3. หาจำนวนผู้ป่วยที่เป็นคน "**ที่ไม่ใช่คนไทย**"

In [28]:
this_data = data_covid[['sex','age','province_of_onset']] 

In [29]:
this_data  #ตัดออกมาจากตารางทั้งหมด จะได้ 3 คอลัมน์

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [30]:
male = this_data [this_data ['sex']=='ชาย'] # เลือกมาเฉพาะค่าที่เป็นเพศชาย
male

sex   age province_of_onset
3      ชาย  68.0     กรุงเทพมหานคร
8      ชาย  63.0     กรุงเทพมหานคร
10     ชาย  33.0     กรุงเทพมหานคร
11     ชาย  61.0     กรุงเทพมหานคร
12     ชาย   6.0     กรุงเทพมหานคร
...    ...   ...               ...
12640  ชาย  66.0   พระนครศรีอยุธยา
12641  ชาย  29.0          ปทุมธานี
12643  ชาย  39.0              สตูล
12646  ชาย  18.0       สมุทรปราการ
12647  ชาย  42.0     กรุงเทพมหานคร

[4819 rows x 3 columns]

In [31]:
male['age'] # หาอายุของเพศชาย

3        68.0
8        63.0
10       33.0
11       61.0
12        6.0
         ... 
12640    66.0
12641    29.0
12643    39.0
12646    18.0
12647    42.0
Name: age, Length: 4819, dtype: float64

In [ ]:
sum = 0 # sum มารับ โดยให้เป็น 0 เนื่องจากจะนำมารับค่าทั้งหมดที่จะบวกกัน
N = 0 #จะนับว่ามีจำนวนคนทั้งหมดกี่คน เริ่มที่ 0
for a in male['age']: # นำค่าอายุของเพศชายเข้ามาเพื่อวนดูค่าอายุ 
    
    if a > 0: # ถ้า a > 0 แสดงว่าเป็นอายุของคน ดังนั้นจะนำมานับ
        sum += a # sum = sum + a  # คำนวณหาอายุของเพศชาย ซึ่งถ้า a เป็นอายุคนก็จะถูกนำมาบวกกันเรื่อยๆ
        N += 1 # คำนวณหาจำนวนผู้ชายทั้งหมดโดยเพิ่มทีละ 1

print(f'อายุเฉลี่ย ของ ผู้ชาย {sum/N}') # แสดงค่าเฉลี่ยของอายุผู้ป่วยชาย โดยให้ sum = อายุรวมทั้งหมดของผู้ป่วยชาย หารด้วย N = จำนวนผู้ป่วยชายทั้งหมด

อายุเฉลี่ย ของ ผู้ชาย 39.46076940331987


In [ ]:
male_KK = male[male['province_of_onset']=='ขอนแก่น'] # เลือกค่าที่เป็นเพศชาย ที่อยู่ในจังหวัดขอนแก่น

sum = 0 # sum มารับ โดยให้เป็น 0 เนื่องจากจะนำมารับค่าทั้งหมดที่จะบวกกัน
N = 0 # N นับจำนวนคนทั้งหมด เริ่มที่ 0
for a in male_KK['age']: # นำค่าอายุของเพศชายเข้ามาเพื่อวนดูค่าอายุ 
    
    if a > 0: # ถ้า a > 0 แสดงว่าเป็นอายุของคน ดังนั้นจะนำมานับ
        sum += a # sum = sum + a  คำนวณหาอายุของเพศชาย ซึ่งถ้า a เป็นอายุคนก็จะถูกนำมาบวกกันเรื่อยๆ
        N += 1 # คำนวณหาจำนวนผู้ชายทั้งหมดโดยเพิ่มทีละ 1

print(f'อายุเฉลี่ย ของผู้ชาย ในจังหวัดขอนแก่น {sum/N}') # แสดงค่าเฉลี่ยของอายุผู้ชายในจังหวัดขอนแก่น โดยให้ sum = อายุรวมของผู้ชายในจังหวัดขอนแก่น หารด้วย N = จำนวนผู้ป่วยชายทั้งหมด

อายุเฉลี่ย ของผู้ชาย ในจังหวัดขอนแก่น 26.571428571428573


In [32]:
dataN = data_covid[['nationality']] # เลือกข้อมูลที่ต้องการนำมาใช้งาน โดยเลือกเฉพาะ column 'nationality' ในการนำมาหาผู้ป่วยที่ไม่ใช่คนไทย
dataN # แสดงค่าข้อมูลว่ามีประเทศอะไรบ้าง

nationality
0           China
1           China
2        Thailand
3           China
4           China
...           ...
12648    Thailand
12649    Thailand
12650    Thailand
12651    Thailand
12652       Burma

[12653 rows x 1 columns]

In [33]:
patient = dataN[dataN['nationality'] != 'Thailand'] #กำหนดค่าตัวแปร patient โดยใช้คอลัมน์ nationality ต้องการหาผู้ป่วยที่ "ไม่ใช่คนไทย"
patient # แสดงข้อมูลผู้ป่วยที่ไม่ใช่คนไทย

nationality
0           China
1           China
3           China
4           China
5           China
...           ...
12614       Burma
12615       Burma
12617     Ireland
12618    Pakistan
12652       Burma

[5340 rows x 1 columns]

In [34]:
# 3.หาจำนวนผู้ป่วยที่เป็นคน "ไม่ใช่คนไทย"

patient = dataN[dataN['nationality'] != 'Thailand'] # หาผู้ป่วยที่ไม่ใช่คนไทย จาก column 'nationality' เลือกโดย ไม่เอาไทย
total = patient.count() # นำค่าที่ไม่มีคนไทยมาหาจำนวนผู้ป่วยทั้งหมด โดยใช้คำสั่ง count ในการนับจำนวน #count จะนับเป็นจำนวนครั้งเมื่อค้นเจอค่าข้อมูล
total # แสดงค่าออกมา

nationality    4129
dtype: int64

จากการคำนวณหาจำนวนผู้ป่วยที่ไม่ใช่คนไทย โดยใช้คำสั่ง count จะเห็นได้ว่า ผู้ป่วยที่ไม่ใช่คนไทยมีจำนวน 4129 คน (ดูจาก column 'nationality')

การจัดการ Missing Value

- ลบ record ที่เป็น missing 
- แทนที่ ค่า missing ด้วยค่าที่เหมาะสม mean, default, category-unknown
- ใช้ ค่าจาก columns อื่นๆ ช่วยประมาณค่า ค่าใน column ที่หายไป (regression, deep learning, etc.)

### ลบ record (dropna)

missing = ```None``` ```NA```(not autorized) ```NaN``` (not a number)

In [55]:
this_data #เลือกตารางมา

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [35]:
this_data.shape #(12423, 3)

(12653, 3)

In [92]:
this_data.dropna() #จาก 12653 เหลือ 8478

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8478 rows x 3 columns]

In [36]:
print(data_covid.shape)
print(data_covid.dropna().shape)

(12653, 10)
(6655, 10)


In [57]:
this_data.dropna(subset = ['age','sex'])

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

In [58]:
print(this_data.dropna().shape)
print(this_data.dropna(subset = ['age','sex']).shape)

(8478, 3)
(8880, 3)


## ตรวจสอบ Missing (```isnull()```)

In [59]:
this_data.isnull()

sex    age  province_of_onset
0      False  False              False
1      False  False              False
2      False  False              False
3      False  False              False
4      False  False              False
...      ...    ...                ...
12648  False  False              False
12649  False  False              False
12650  False  False              False
12651  False  False              False
12652  False  False              False

[12653 rows x 3 columns]

In [60]:
this_data.isnull().any() # .any() เอาค่าความจริงในคอลัมน์ มา OR กัน

sex                  True
age                  True
province_of_onset    True
dtype: bool

In [61]:
this_data.isnull().all() # .all() เอาค่าความจริงในคอลัมน์ มา AND กัน

sex                  False
age                  False
province_of_onset    False
dtype: bool

In [62]:
this_data.isnull().any().any()

True

In [65]:
this_data.dropna(subset = ['sex','age']).isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [66]:
this_data.isnull().any()

sex                  True
age                  True
province_of_onset    True
dtype: bool

In [67]:
this_data.dropna(subset = ['sex','age'])

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

### 1. เพื่อให้ได้ผลลัพธ์การจัดการ (```dropna()```) ต้องเอาตัวแปรมารับ

In [68]:
this_data_nonull = this_data.dropna(subset = ['sex','age'])

In [71]:
this_data_nonull.isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

### 2. update ตารางด้วยตัวแปร (```inplace```)

In [94]:
this_data.dropna(subset=['sex','age'],inplace=True) #เตือนว่าจะเกิดข้อผิดพลาด

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [95]:
this_data.isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [75]:
aa=this_data.dropna(subset = ['sex','age'],inplace=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [76]:
aa

### แทน missing ด้วยค่าที่เหมาะสม (```fillna```)

In [45]:
this_data = data_covid[['sex','age','province_of_onset']]

In [97]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

In [98]:
print(f'จำนวน record ก่อน drop missing ใน province {this_data.shape[0]}')
print(f'จำนวน record หลัง drop missing ใน province {this_data.dropna(subset=["province_of_onset"]).shape[0]}')

จำนวน record ก่อน drop missing ใน province 8880
จำนวน record หลัง drop missing ใน province 8478


### 1. แทนด้วย class ใหม่

In [99]:
this_data_updated = this_data.fillna(value={'sex':'ไม่รู้', 'age':-99, 'province_of_onset':'ไม่รู้'})
this_data_updated.head()

sex   age province_of_onset
0  หญิง  61.0     กรุงเทพมหานคร
1  หญิง  74.0     กรุงเทพมหานคร
2  หญิง  73.0            นครปฐม
3   ชาย  68.0     กรุงเทพมหานคร
4  หญิง  66.0     กรุงเทพมหานคร

In [116]:
this_data_updated.tail(15)

sex   age province_of_onset
12633  หญิง  44.0         สมุทรสาคร
12634  หญิง  54.0         สมุทรสาคร
12640   ชาย  66.0   พระนครศรีอยุธยา
12641   ชาย  29.0          ปทุมธานี
12642  หญิง  50.0           ราชบุรี
12643   ชาย  39.0              สตูล
12644  หญิง  49.0       สมุทรปราการ
12645  หญิง  44.0            ชลบุรี
12646   ชาย  18.0       สมุทรปราการ
12647   ชาย  42.0     กรุงเทพมหานคร
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

In [101]:
this_data_updated.isnull().any()

sex                  False
age                  False
province_of_onset    False
dtype: bool

### 2. แทนด้วยค่าที่เหมาะสม


*   ตามใจคนทำ
*   ค่ากลาง


In [83]:
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [102]:
data_covid_fillasiwant = data_covid.fillna(value={'sex':'หญิง'})  #เรามีข้อมูลว่า เพศหญิงมีเยอะกว่าเพศชาย

In [103]:
data_covid_fillasiwant.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                      False
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

#### แทนด้วยค่ากลาง

In [104]:
data_covid['age']

0        61.0
1        74.0
2        73.0
3        68.0
4        66.0
         ... 
12648    44.0
12649    52.0
12650    23.0
12651    29.0
12652    22.0
Name: age, Length: 12653, dtype: float64

In [105]:
data_covid['age'].mean()

38.054748339151

In [106]:
data_covid_fillwithmean = data_covid.fillna(value={'age':data_covid['age'].mean()})

In [107]:
data_covid_fillwithmean.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                       True
age                      False
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

### การวนลูป record ในตาราง  (```.iterrows```)

In [108]:
this_data = data_covid[['sex','age','province_of_onset']]
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [109]:
for each_row in this_data.iterrows():
    # print(each_row)
    # print(each_row[1])
    # print(each_row[1]['sex'])
    if (each_row[1]['age'] <= 20) and (each_row[1]['province_of_onset'] == 'ขอนแก่น'):
        print(each_row)

(1466, sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object)
(9333, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object)
(9334, sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object)
(10610, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object)
(11517, sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object)
(11697, sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object)


### การวนลูป แบบมองตารางแพนด้าส์(pandas dataframe) เป็น numpy array  (```.iloc```)

In [110]:
for each_row in range(this_data.shape[0]):
    if (this_data.iloc[each_row,1] <= 20) and (this_data.iloc[each_row,2] == 'ขอนแก่น'):  #each_row คือ ตัวเลข
        print(each_row)
        print(this_data.iloc[each_row,:])   

1466
sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object
9333
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object
9334
sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object
10610
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object
11517
sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object
11697
sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object


## Quiz ตัดตารางออกมาเฉพาะปี 2021 ```announce_date``` ในปี 2021 

**Hint**
*   วนลูปหา index ของปี 2021
*   ตัดตารางมาเฉพาะ ปี 2021

In [111]:
TF=list()
for each_row in data_covid.iterrows():
    if each_row[1]['announce_date'].split('/')[2] == '2021':
        TF.append(True)
    else:
        TF.append(False)

data_covid[TF].head()

No.  ...                                               risk
6884  6885  ...                                  Cluster สมุทรสาคร
6885  6886  ...                                      Cluster ระยอง
6886  6887  ...                                        สถานบันเทิง
6887  6888  ...                                  Cluster สมุทรสาคร
6888  6889  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...

[5 rows x 10 columns]

### HW กลุ่ม 2 || เลือกข้อมูลมาเฉพาะวัน announce_date เป็นวันที่ 1-20  

# Function ตัวช่วยใน pandas

### ```.describe()``` คำนวณค่าทางสถิติของข้อมูลที่เป็นตัวเลข

In [113]:
data_covid.describe()

No.          age
count  12653.000000  8881.000000
mean    6327.000000    38.054748
std     3652.750813    15.125178
min        1.000000     0.100000
25%     3164.000000    27.000000
50%     6327.000000    36.000000
75%     9490.000000    48.000000
max    12653.000000    97.000000

### ```.mean()``` คำนวณค่าเฉลี่ยของข้อมูลโดยไม่สนใจ missing 

In [114]:
data_covid[data_covid['sex']=='ชาย']['age'].mean()

39.46076940331987

### ```.isnull()```

In [115]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
12648  False          False  ...              False  False
12649  False          False  ...              False  False
12650  False          False  ...              False  False
12651  False          False  ...              False  False
12652  False          False  ...              False  False

[12653 rows x 10 columns]

# next

*   ต่อตารางแกน X แกน y
*   .groupby
*   save table